scipy/optimize/_qap.py

```diff
@@ -0,0 +1,784 @@
+           C.E. Priebe, "Fast approximate quadratic programming for graph
+           matching," PLOS one, vol. 10, no. 4, p. e0121002, 2015,
+           https://doi.org/10.1371/journal.pone.0121002
+
+    .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,
+           C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
+           203-215, https://doi.org/10.1016/j.patcog.2018.09.014
+    """
+
+    _check_unknown_options(unknown_options)
+
+    maxiter = operator.index(maxiter)
+
+    # ValueError check
+    A, B, partial_match = _common_input_validation(A, B, partial_match)
+
+    msg = None
+    if (A < 0).any() or (B < 0).any():
+        msg = "`A` and `B` matrices must contain only non-negative elements."
+    elif isinstance(init_J, str) and init_J not in {'barycenter'}:
+        msg = "Invalid 'init_J' parameter string"
+    elif init_weight < 0 or init_weight > 1:
+        msg = "'init_weight' must be strictly between zero and one"
+    elif maxiter <= 0:
+        msg = "'maxiter' must be a positive integer"
+    elif tol <= 0:
+        msg = "'tol' must be a positive float"
+    if msg is not None:
+        raise ValueError(msg)
+
+    rng = check_random_state(rng)
+    n = A.shape[0]  # number of vertices in graphs
+    n_seeds = partial_match.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+
+    nonseed_B = np.setdiff1d(range(n), partial_match[:, 1])
+    if shuffle_input:
+        nonseed_B = rng.permutation(nonseed_B)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    nonseed_A = np.setdiff1d(range(n), partial_match[:, 0])
+    perm_A = np.concatenate([partial_match[:, 0],
+                             nonseed_A], axis=None).astype(int)
+    perm_B = np.concatenate([partial_match[:, 1],
+                             nonseed_B], axis=None).astype(int)
+
+    # definitions according to Seeded Graph Matching [2].
+    A11, A12, A21, A22 = _split_matrix(A[perm_A][:, perm_A], n_seeds)
+    B11, B12, B21, B22 = _split_matrix(B[perm_B][:, perm_B], n_seeds)
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    if init_weight != 1:
+        # generate a nxn matrix where each entry is a random number [0, 1]
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+
+    # OPTIMIZATION WHILE LOOP BEGINS
```

Comment:
A comment is not inappropriate here, although it could be more descriptive to help the reader get their bearings relative to the paper.
```suggestion
    # Algorithm 1 Step 2 (main loop)
```